File: cpp/problem-003.cpp
#include <vector>
#include <cmath>
#include <algorithm>

#include "benchmark/timer.hpp"

std::vector<uint32_t> findPrimeFactors(uint64_t number) {
    std::vector<uint32_t> factors;

    for (size_t i = 2; i <= sqrt(number); ++i) {
        while (number % i == 0) {
            factors.push_back(i);
            number /= i;
        }
    }

    if (number > 2)
        factors.push_back(number);

    return factors;
}

void solve() {
    uint64_t THE_NUMBER = 600851475143LL;

    auto primeFactors = findPrimeFactors(THE_NUMBER);

    auto maxElement = std::max_element(primeFactors.begin(), primeFactors.end());

    std::cout << "Problem 3: " << *maxElement << std::endl;
}

int main() {
    Timer timer;
    solve();
}

File: python/problem003.py
def prime_factors(n: int) -> list[int]:
    factors = []
    d = 2
    while d * d <= n:
        while (n % d) == 0:
            factors.append(d)
            n //= d
        d += 1
    if n > 1:
        factors.append(n)
    return factors


def solve() -> int:
    n = 600851475143
    factors = prime_factors(n)
    return max(factors)


if __name__ == '__main__':
    print(solve())
